### PR TITLE
[REEF-514] Move Driver restart configuration options to a separate configuration

### PR DESCRIPTION
**Pull request by afchung**
This addressed the issue by
  * Moving configurations for restarts to a separate ConfigurationModule.
  * Refactoring restart handlers to restart specific configuration modules that are runtime-dependent.

JIRA:
  [REEF-514](https://issues.apache.org/jira/browse/REEF-514)

### CODE REVIEW

**Comment by afchung** [comment]
This is precursory work of REEF-483.

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobClient.java [hunk 1]
```diff
@@ -139,17 +137,29 @@ public static Configuration getHTTPConfiguration() {
...
-            ReefEventStateManager.DriverRestartTaskRunningStateHandler.class)
         .set(DriverServiceConfiguration.ON_DRIVER_STARTED, ReefEventStateManager.StartStateHandler.class)
         .set(DriverServiceConfiguration.ON_DRIVER_STOP, ReefEventStateManager.StopStateHandler.class)
         .build();
+
     return Configurations.merge(httpHandlerConfiguration, driverConfigurationForHttpServer);
   }
 
+  public static Configuration getYarnConfiguration() {
+    Configuration yarnDriverRestartConfiguration = YarnDriverRestartConfiguration.CONF
```
Comment (markusweimer) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobClient.java]: `final`

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobClient.java [hunk 2]
```diff
@@ -139,17 +137,29 @@ public static Configuration getHTTPConfiguration() {
...
+        .set(YarnDriverRestartConfiguration.ON_DRIVER_RESTARTED, ReefEventStateManager.DriverRestartHandler.class)
+        .build();
+
+    Configuration driverRestartHandlerConfigurations = DriverRestartHandlersConfiguration.CONF
```
Comment (markusweimer) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/generic/JobClient.java]: `final`

File: lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import org.apache.reef.driver.parameters.DriverRestartTaskRunningHandlers;
+import org.apache.reef.driver.task.RunningTask;
+import org.apache.reef.runtime.common.DriverRestartCompleted;
+import org.apache.reef.tang.formats.ConfigurationModule;
+import org.apache.reef.tang.formats.ConfigurationModuleBuilder;
+import org.apache.reef.tang.formats.OptionalImpl;
+import org.apache.reef.wake.EventHandler;
+import org.apache.reef.wake.time.event.StartTime;
+
+public final class DriverRestartHandlersConfiguration extends ConfigurationModuleBuilder {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java]: Why not have all of this in `DriverRestartConfiguration`?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java [hunk 4]
```diff
@@ -59,11 +59,11 @@
   private static final Logger LOG = Logger.getLogger(REEFLauncher.class.getName());
 
   private static final Configuration LAUNCHER_STATIC_CONFIG = Tang.Factory.getTang().newConfigurationBuilder()
-          .bindSetEntry(Clock.StartHandler.class, PIDStoreStartHandler.class)
-          .bindNamedParameter(RemoteConfiguration.ErrorHandler.class, REEFErrorHandler.class)
-          .bindNamedParameter(RemoteConfiguration.ManagerName.class, "REEF_LAUNCHER")
-          .bindNamedParameter(RemoteConfiguration.MessageCodec.class, REEFMessageCodec.class)
-          .build();
+      .bindSetEntry(Clock.StartHandler.class, PIDStoreStartHandler.class)
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java]: This is purely cosmetic, right? If so, please remove the change.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java [hunk 5]
```diff
@@ -77,16 +77,16 @@
 
   @Inject
   private REEFLauncher(@Parameter(ClockConfigurationPath.class) final String configurationPath,
-               @Parameter(ProfilingEnabled.class) final boolean enableProfiling,
-               final ConfigurationSerializer configurationSerializer,
-               final REEFVersion reefVersion) {
+                       @Parameter(ProfilingEnabled.class) final boolean enableProfiling,
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java]: Same as above: If this is purely cosmetic, please remove the change.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java [hunk 6]
```diff
@@ -77,16 +77,16 @@
...
+                       final ConfigurationSerializer configurationSerializer,
+                       final REEFVersion reefVersion) {
     this.configurationPath = configurationPath;
     this.configurationSerializer = configurationSerializer;
     this.isProfilingEnabled = enableProfiling;
     this.reefVersion = reefVersion;
     this.clockConfig = Configurations.merge(
-            readConfigurationFromDisk(this.configurationPath, this.configurationSerializer),
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java]: Cosmetic?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java [hunk 7]
```diff
@@ -118,7 +118,7 @@ private static RuntimeException fatal(final REEFErrorHandler errorHandler, final
   }
 
   private static Configuration readConfigurationFromDisk(
-          final String configPath, final ConfigurationSerializer serializer) {
+      final String configPath, final ConfigurationSerializer serializer) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java]: Cosmetic?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java [hunk 8]
```diff
@@ -152,7 +152,7 @@ public static void main(final String[] args) {
     LOG.log(Level.FINE, "REEFLauncher started with user name [{0}]", System.getProperty("user.name"));
 
     LOG.log(Level.FINE, "REEFLauncher started. Assertions are {0} in this process.",
-            EnvironmentUtils.areAssertionsEnabled() ? "ENABLED" : "DISABLED");
+        EnvironmentUtils.areAssertionsEnabled() ? "ENABLED" : "DISABLED");
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java]: Cosmetic?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java [hunk 9]
```diff
@@ -195,7 +195,7 @@ public static void main(final String[] args) {
    * @param propNames property names.
    */
   public static void propagateProperties(
-          final Collection<String> vargs, final boolean copyNull, final String... propNames) {
+      final Collection<String> vargs, final boolean copyNull, final String... propNames) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java]: Cosmetic?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java [hunk 10]
```diff
@@ -212,7 +212,7 @@ public static void propagateProperties(
    * Same as above, but with copyNull == false by default.
    */
   public static void propagateProperties(
-          final Collection<String> vargs, final String... propNames) {
+      final Collection<String> vargs, final String... propNames) {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFLauncher.java]: Cosmetic?

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloDriver.java [hunk 11]
```diff
@@ -68,6 +68,31 @@ public void onNext(final StartTime startTime) {
   }
 
   /**
+   * Handles the StartTime event: Fails intentionally.
+   */
+  public final class FailureToStartHandler implements EventHandler<StartTime> {
```
Comment (markusweimer) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloDriver.java]: This shouldn't be in the simple example. Maybe we can turn this into the nucleus of the integration test we'll need eventually?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnDriverConfiguration.java [hunk 12]
```diff
@@ -29,7 +28,6 @@
  * Additional YARN-Specific configuration options to be merged with DriverConfiguration.
  */
 @Public
-@ClientSide
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnDriverConfiguration.java]: Why this change?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java [hunk 13]
```diff
@@ -392,7 +392,7 @@ private void onContainerStatus(final ContainerStatus value) {
           status.setState(ReefServiceProtos.State.FAILED);
         }
         status.setExitCode(value.getExitStatus());
-          // remove the completed container (can be either done/killed/failed) from book keeping
+        // remove the completed container (can be either done/killed/failed) from book keeping
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java]: Cosmetic?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java [hunk 14]
```diff
@@ -482,7 +482,7 @@ private synchronized void doHomogeneousRequests() {
       final AMRMClient.ContainerRequest firstRequest = this.requestsBeforeSentToRM.peek();
 
       while (!this.requestsBeforeSentToRM.isEmpty() &&
-             isSameKindOfRequest(firstRequest, this.requestsBeforeSentToRM.peek())) {
+          isSameKindOfRequest(firstRequest, this.requestsBeforeSentToRM.peek())) {
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnContainerManager.java]: Cosmetic?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverConfiguration.java [hunk 15]
```diff
@@ -35,10 +35,7 @@
 import org.apache.reef.runtime.yarn.driver.parameters.JobSubmissionDirectory;
 import org.apache.reef.runtime.yarn.driver.parameters.YarnHeartbeatPeriod;
 import org.apache.reef.runtime.yarn.util.YarnConfigurationConstructor;
-import org.apache.reef.tang.formats.ConfigurationModule;
-import org.apache.reef.tang.formats.ConfigurationModuleBuilder;
-import org.apache.reef.tang.formats.OptionalParameter;
-import org.apache.reef.tang.formats.RequiredParameter;
+import org.apache.reef.tang.formats.*;
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverConfiguration.java]: Mostly a cosmetic change, please undo.

**Comment by markusweimer** [comment]
@afchung I did a pass. Mostly cosmetic (un-)changes.

File: lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java (hunk 3 above)
Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java]: DriverRestartConfiguration cannot be bound multiple times since not all of its members are of Set<>.

Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java]: So if a user wants to add more than one specific type of RestartContextActiveHandler, for instance, they wouldn't be able to do it just through DriverRestartConfig.

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloDriver.java (hunk 11 above)
Comment (afchung) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloDriver.java]: Removed.

**Comment by afchung** [comment]
@markusweimer Addressed your comments, please have another pass. Thanks!

**Comment by markusweimer** [comment]
@afchung Thanks, I will do another pass now.

File: lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java (hunk 3 above)
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java]: Ah, now I understand the confusion. 

If the underlying named parameter is a `Set<>`, user code is free to call `ConfigurationModule.set()` many times. So, if they want to bind more than one `RestartContextActiveHandler`, they could.

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloDriver.java [hunk 16]
```diff
@@ -68,6 +68,21 @@ public void onNext(final StartTime startTime) {
   }
 
   /**
+   * Handles the StartTime event: Request as single Evaluator.
+   */
+  public final class RestartHandler implements EventHandler<StartTime> {
```
Comment (markusweimer) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloDriver.java]: Can we move this to a different example? E.g. HelloWithRestart or such? I fear that its presence will confuse the audience of the `HelloREEF` example, newbies to REEF.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverConfiguration.java [hunk 17]
```diff
@@ -74,12 +74,12 @@
    */
   public static final OptionalParameter<Double> JVM_HEAP_SLACK = new OptionalParameter<>();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverConfiguration.java]: All changes in this file are cosmetic, please remove.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRestartConfiguration.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+ * <p/>
+ */
+@ClientSide
+@Public
+@Provided
+public final class YarnDriverRestartConfiguration extends ConfigurationModuleBuilder {
+  /**
+   * This event is fired in place of the ON_DRIVER_STARTED when the Driver is in fact restarted after failure.
+   */
+  public static final RequiredImpl<EventHandler<StartTime>> ON_DRIVER_RESTARTED = new RequiredImpl<>();
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRestartConfiguration.java]: This would be a common event type across all runtimes that support restarts, right? If so, shouldn't it be available on `DriverRestartConfiguration`?

**Comment by markusweimer** [comment]
@afchung Done with my pass.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRestartConfiguration.java (hunk 18 above)
Comment (afchung) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRestartConfiguration.java]: My original intention here is to force the user to bind at least one RestartHandler, since in my first PR for REEF-483, I was using the existence of a RestartHandler to determine whether or not the user wants to restart their job. On second thought, I should probably be using the existence of the runtime RestartManager (to be implemented in REEF-483) to determine that instead, Will remove.

Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRestartConfiguration.java]: You can also use the *existence* of a `DriverRestartHandler` to distinguish.

File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloDriver.java (hunk 16 above)
Comment (afchung) [inline lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloDriver.java]: OK, I'll create a new class for a new example in my next PR for REEF-483. Removed.

File: lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java (hunk 3 above)
Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java]: OK, the usage of the term set mislead me. Will merge into the other configuration. Thanks!

Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java]: Yes, it is a bit wonky. The reason for this is historic: Originally, Tang didn't support `Set<T>`, and REEF only allowed *one* handler per event type. When we generalized from that, we wanted to maintain backwards compatibility. Hence, we did *not* add a new method `add(...)` or such, but just overloaded `set(...)`. That way, application code already written didn't have to change.

Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/client/parameters/DriverRestartHandlersConfiguration.java]: @markusweimer Actually, on second thought, I will be renaming DriverRestartConfiguration to DriverRuntimeRestartConfiguration, since what will eventually be bound here are EventHandlers that depend on the runtime DriverRestartConfiguration (e.g. DriverRestartManager in YarnDriverRestartConfiguration). I will also be renaming DriverRestartHandlersConfiguration to DriverRestartConfiguration, make it Public and ClientSide.

**Comment by afchung** [comment]
@markusweimer Updated and rebased. Please have another look. Note that the original DriverRestartConfiguration has been renamed and repurposed to DriverRuntimeRestartConfiguration, since it won't be exposed to the user. DriverRuntimeRestartConfiguration does not contain anything meaningful yet, but it will be used to bind the runtime-dependent handlers in REEF-483. The original DriverRestartHandlersConfiguration has been renamed to DriverRestartConfiguration and will be the  configuration module exposed to the user. Thanks!

**Comment by markusweimer** [comment]
@afchung Thanks! I will now do the next pass.

**Comment by markusweimer** [comment]
This looks good. I will test on my dev box and merge.

**Comment by markusweimer** [comment]
tests pass on my windows machine and the linux build server. merging.
